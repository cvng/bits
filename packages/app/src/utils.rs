#![forbid(unsafe_code)]

use askama::*;
use http::StatusCode;
use poem::IntoResponse;
use poem::Response;

pub fn into_response<T: Template>(t: &T) -> Response {
  match t.render() {
    Ok(body) => Response::builder()
      .header(
        http::header::CONTENT_TYPE,
        http::HeaderValue::from_static(T::MIME_TYPE),
      )
      .body(body)
      .into_response(),
    Err(_) => StatusCode::INTERNAL_SERVER_ERROR.into_response(),
  }
}
