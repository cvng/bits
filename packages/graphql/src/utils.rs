use async_graphql::Request;
use async_graphql::Variables;
use graphql_client::QueryBody;
use serde::Serialize;
use serde_json::to_value;

pub fn try_into_request<V>(
  query_body: QueryBody<V>,
) -> Result<Request, serde_json::Error>
where
  V: Serialize,
{
  let request = Request::new(query_body.query)
    .operation_name(query_body.operation_name)
    .variables(Variables::from_json(to_value(query_body.variables)?));

  Ok(request)
}
