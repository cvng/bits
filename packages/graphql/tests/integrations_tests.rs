mod utils;

use crate::utils::execute;
use crate::utils::Context;
use crate::utils::TestToken;
use graphql_client::GraphQLQuery;
#[cfg(test)]
use insta::assert_json_snapshot;
use test_context::test_context;
use tokio::test;

#[rustfmt::skip]
#[derive(graphql_client::GraphQLQuery)]
#[graphql(schema_path = "../../docs/schema.gql", query_path = "tests/operations.graphql")]
pub struct CreateShowMutation;

#[test_context(Context)]
#[test]
async fn test_create_show_mutation(ctx: &mut Context) {
  let response = execute(
    ctx,
    CreateShowMutation::build_query(create_show_mutation::Variables {}),
    Some(TestToken::seller_token()),
  )
  .await
  .unwrap();

  assert_json_snapshot!(response, { ".data.createShow.show.id" => "[uuid]" });
}

#[rustfmt::skip]
#[derive(GraphQLQuery)]
#[graphql(schema_path = "../../docs/schema.gql", query_path = "tests/operations.graphql")]
pub struct CreateProductMutation;

#[test_context(Context)]
#[test]
async fn test_create_product_mutation(ctx: &mut Context) {
  let response = execute(
    ctx,
    CreateProductMutation::build_query(create_product_mutation::Variables {}),
    Some(TestToken::seller_token()),
  )
  .await
  .unwrap();

  assert_json_snapshot!(response, { ".data.createProduct.product.id" => "[uuid]" });
}

#[rustfmt::skip]
#[derive(GraphQLQuery)]
#[graphql(schema_path = "../../docs/schema.gql", query_path = "tests/operations.graphql")]
pub struct CreateAuctionMutation;

#[test_context(Context)]
#[test]
async fn test_create_auction_mutation(ctx: &mut Context) {
  let response = execute(
    ctx,
    CreateAuctionMutation::build_query(create_auction_mutation::Variables {}),
    Some(TestToken::seller_token()),
  )
  .await
  .unwrap();

  assert_json_snapshot!(response, { ".data.createAuction.auction.id" => "[uuid]" });
}

#[rustfmt::skip]
#[derive(GraphQLQuery)]
#[graphql(schema_path = "../../docs/schema.gql", query_path = "tests/operations.graphql")]
pub struct StartMutation;

#[test_context(Context)]
#[test]
async fn test_start_mutation(ctx: &mut Context) {
  let response = execute(
    ctx,
    StartMutation::build_query(start_mutation::Variables {}),
    Some(TestToken::seller_token()),
  )
  .await
  .unwrap();

  assert_json_snapshot!(response, { ".data.start.auction.id" => "[uuid]" });
}

#[rustfmt::skip]
#[derive(GraphQLQuery)]
#[graphql(schema_path = "../../docs/schema.gql", query_path = "tests/operations.graphql")]
pub struct CommentMutation;

#[test_context(Context)]
#[test]
async fn test_comment_mutation(ctx: &mut Context) {
  let response = execute(
    ctx,
    CommentMutation::build_query(comment_mutation::Variables {}),
    Some(TestToken::buyer_token()),
  )
  .await
  .unwrap();

  assert_json_snapshot!(response, { ".data.comment.comment.id" => "[uuid]" });
}

#[rustfmt::skip]
#[derive(GraphQLQuery)]
#[graphql(schema_path = "../../docs/schema.gql", query_path = "tests/operations.graphql")]
pub struct BidMutation;

#[test_context(Context)]
#[test]
async fn test_bid_mutation(ctx: &mut Context) {
  let response = execute(
    ctx,
    BidMutation::build_query(bid_mutation::Variables {}),
    Some(TestToken::buyer_token()),
  )
  .await
  .unwrap();

  assert_json_snapshot!(response, { ".data.bid.bid.id" => "[uuid]" });
}
