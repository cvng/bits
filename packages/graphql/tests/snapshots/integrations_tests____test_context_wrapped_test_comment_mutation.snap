---
source: packages/graphql/tests/integrations_tests.rs
expression: response
---
{
  "data": {
    "comment": {
      "comment": {
        "id": "[uuid]",
        "authorId": "00000000-2000-0000-0000-000000000000",
        "showId": "00000000-0000-0000-0000-000000000000",
        "text": "This is a comment"
      }
    }
  }
}
